File: ta4j-core/src/test/java/org/ta4j/core/rules/XorRuleTest.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.rules;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.ta4j.core.BarSeries;
import org.ta4j.core.Rule;
import org.ta4j.core.mocks.MockBarSeriesBuilder;

public class XorRuleTest {

    private Rule satisfiedRule;
    private Rule unsatisfiedRule;
    private BarSeries series;
    private RuleTraceTestLogger ruleTraceTestLogger;

    @Before
    public void setUp() {
        ruleTraceTestLogger = new RuleTraceTestLogger();
        ruleTraceTestLogger.open();

        satisfiedRule = new BooleanRule(true);
        unsatisfiedRule = new BooleanRule(false);
        series = new MockBarSeriesBuilder().withData(1).build();
    }

    @After
    public void tearDownLogger() {
        ruleTraceTestLogger.close();
    }

    @Test
    public void isSatisfied() {
        assertTrue(satisfiedRule.xor(BooleanRule.FALSE).isSatisfied(0));
        assertTrue(BooleanRule.FALSE.xor(satisfiedRule).isSatisfied(0));
        assertFalse(unsatisfiedRule.xor(BooleanRule.FALSE).isSatisfied(0));
        assertFalse(BooleanRule.FALSE.xor(unsatisfiedRule).isSatisfied(0));

        assertFalse(satisfiedRule.xor(BooleanRule.TRUE).isSatisfied(10));
        assertFalse(BooleanRule.TRUE.xor(satisfiedRule).isSatisfied(10));
        assertTrue(unsatisfiedRule.xor(BooleanRule.TRUE).isSatisfied(10));
        assertTrue(BooleanRule.TRUE.xor(unsatisfiedRule).isSatisfied(10));
    }

    @Test
    public void serializeAndDeserialize() {
        Rule xor = satisfiedRule.xor(unsatisfiedRule);
        RuleSerializationRoundTripTestSupport.assertRuleRoundTrips(series, xor);
        RuleSerializationRoundTripTestSupport.assertRuleJsonRoundTrips(series, xor);
    }

    @Test
    public void traceLoggingVerboseModePreservesBothChildRuleLogs() {
        FixedRule rule1 = new FixedRule(1);
        rule1.setName("Xor Rule 1");
        FixedRule rule2 = new FixedRule(2);
        rule2.setName("Xor Rule 2");
        XorRule xorRule = new XorRule(rule1, rule2);
        xorRule.setName("Xor Pair");
        xorRule.setTraceMode(Rule.TraceMode.VERBOSE);

        ruleTraceTestLogger.clear();
        xorRule.isSatisfied(1);

        String logContent = ruleTraceTestLogger.getLogOutput();
        assertTrue("Verbose mode should log first XOR child", logContent.contains("Xor Rule 1#isSatisfied"));
        assertTrue("Verbose mode should log second XOR child", logContent.contains("Xor Rule 2#isSatisfied"));
        assertTrue("Verbose mode should log parent XOR", logContent.contains("Xor Pair#isSatisfied"));
        assertTrue("Verbose mode should attribute first child path", logContent.contains("path=root.rule1 depth=1"));
        assertTrue("Verbose mode should attribute second child path", logContent.contains("path=root.rule2 depth=1"));
    }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/rules/XorRuleTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.rules

import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.ta4j.core.Rule

class XorRuleTest {
    private lateinit var satisfiedRule: Rule
    private lateinit var unsatisfiedRule: Rule
    @Before
    fun setUp() {
        satisfiedRule = BooleanRule(true)
        unsatisfiedRule = BooleanRule(false)
    }

    @Test
    fun isSatisfied() {
            Assert.assertTrue(satisfiedRule.xor(BooleanRule.FALSE).isSatisfied(0))
            Assert.assertTrue(BooleanRule.FALSE.xor(satisfiedRule).isSatisfied(0))
            Assert.assertFalse(unsatisfiedRule.xor(BooleanRule.FALSE).isSatisfied(0))
            Assert.assertFalse(BooleanRule.FALSE.xor(unsatisfiedRule).isSatisfied(0))
            Assert.assertFalse(satisfiedRule.xor(BooleanRule.TRUE).isSatisfied(10))
            Assert.assertFalse(BooleanRule.TRUE.xor(satisfiedRule).isSatisfied(10))
            Assert.assertTrue(unsatisfiedRule.xor(BooleanRule.TRUE).isSatisfied(10))
            Assert.assertTrue(BooleanRule.TRUE.xor(unsatisfiedRule).isSatisfied(10))
        }
}